File: backend/api_client.ts
/**
 * TypeScript API Client for Corner League Media Backend
 *
 * Provides a fully typed client for the backend API with support for
 * enhanced search functionality, caching, and error handling.
 */

// Types based on backend schemas
export interface TeamResponse {
  id: string;
  sport_id: string;
  league_id?: string;
  name: string;
  market: string;
  slug: string;
  abbreviation?: string;
  logo_url?: string;
  primary_color?: string;
  secondary_color?: string;
  is_active: boolean;
  external_id?: string;
  official_name?: string;
  short_name?: string;
  country_code?: string;
  founding_year?: number;
  sport_name?: string;
  league_name?: string;
  display_name?: string;
  computed_short_name?: string;
  computed_official_name?: string;
  leagues: LeagueInfo[];
}

export interface LeagueInfo {
  id: string;
  name: string;
  slug: string;
  country_code?: string;
  league_level?: number;
  competition_type?: string;
  is_primary: boolean;
  season_start_year?: number;
  position_last_season?: number;
}

export interface SearchMatchInfo {
  field: string;
  value: string;
  highlighted: string;
}

export interface EnhancedTeamResponse extends TeamResponse {
  search_matches: SearchMatchInfo[];
  relevance_score?: number;
}

export interface SearchMetadata {
  query?: string;
  total_matches: number;
  response_time_ms: number;
  filters_applied: Record<string, any>;
  timestamp: string;
}

export interface TeamsPaginatedResponse {
  items: TeamResponse[];
  total: number;
  page: number;
  page_size: number;
  has_next: boolean;
  has_previous: boolean;
}

export interface EnhancedTeamsPaginatedResponse {
  items: EnhancedTeamResponse[];
  total: number;
  page: number;
  page_size: number;
  has_next: boolean;
  has_previous: boolean;
  search_metadata: SearchMetadata;
}

export interface TeamSearchSuggestion {
  suggestion: string;
  type: 'team_name' | 'market' | 'abbreviation';
  team_count: number;
  preview_teams: string[];
}

export interface SearchSuggestionsResponse {
  query: string;
  suggestions: TeamSearchSuggestion[];
  response_time_ms: number;
}

export interface SportResponse {
  id: string;
  name: string;
  slug: string;
  has_teams: boolean;
  icon?: string;
  is_active: boolean;
  display_order: number;
  leagues_count?: number;
}

export interface TeamSearchParams {
  query?: string;
  sport_id?: string;
  league_id?: string;
  market?: string;
  is_active?: boolean;
  page?: number;
  page_size?: number;
}

// Error classes
export class ApiClientError extends Error {
  constructor(message: string, public statusCode?: number) {
    super(message);
    this.name = 'ApiClientError';
  }
}

export class ApiClientTimeoutError extends ApiClientError {
  constructor(message: string) {
    super(message);
    this.name = 'ApiClientTimeoutError';
  }
}

export class ApiClientValidationError extends ApiClientError {
  constructor(message: string) {
    super(message);
    this.name = 'ApiClientValidationError';
  }
}

// Simple cache implementation
class SearchCache {
  private cache = new Map<string, { data: any; expires: number }>();

  constructor(private ttlMs: number = 300000) {} // 5 minutes default

  private getCacheKey(endpoint: string, params: Record<string, any>): string {
    const paramStr = Object.entries(params)
      .filter(([_, value]) => value !== null && value !== undefined)
      .sort(([a], [b]) => a.localeCompare(b))
      .map(([key, value]) => `${key}=${value}`)
      .join('&');
    return `${endpoint}?${paramStr}`;
  }

  get(endpoint: string, params: Record<string, any>): any | null {
    const key = this.getCacheKey(endpoint, params);
    const cached = this.cache.get(key);

    if (cached && Date.now() < cached.expires) {
      return cached.data;
    }

    if (cached) {
      this.cache.delete(key);
    }

    return null;
  }

  set(endpoint: string, params: Record<string, any>, data: any): void {
    const key = this.getCacheKey(endpoint, params);
    this.cache.set(key, {
      data,
      expires: Date.now() + this.ttlMs
    });
  }

  clear(): void {
    this.cache.clear();
  }

  getStats(): { cachedItems: number; ttlMs: number } {
    return {
      cachedItems: this.cache.size,
      ttlMs: this.ttlMs
    };
  }
}

export interface ApiClientOptions {
  baseUrl?: string;
  timeout?: number;
  cacheTtlMs?: number;
  maxRetries?: number;
}

/**
 * Corner League Media API Client
 *
 * Features:
 * - Fully typed methods with validation
 * - Request caching for performance
 * - Automatic retry with exponential backoff
 * - Comprehensive error handling
 * - Search debouncing support
 */
export class CornerLeagueApiClient {
  private baseUrl: string;
  private timeout: number;
  private maxRetries: number;
  private cache: SearchCache;
  private abortController?: AbortController;
  private lastSearchPromise?: Promise<any>;

  constructor(options: ApiClientOptions = {}) {
    this.baseUrl = (options.baseUrl || 'http://127.0.0.1:8001').replace(/\/$/, '');
    this.timeout = options.timeout || 30000;
    this.maxRetries = options.maxRetries || 3;
    this.cache = new SearchCache(options.cacheTtlMs);
  }

  private async request<T>(
    method: string,
    endpoint: string,
    params?: Record<string, any>
  ): Promise<T> {
    const url = new URL(`${this.baseUrl}${endpoint}`);

    if (params) {
      Object.entries(params).forEach(([key, value]) => {
        if (value !== null && value !== undefined) {
          url.searchParams.append(key, String(value));
        }
      });
    }

    let lastError: Error;

    for (let attempt = 0; attempt <= this.maxRetries; attempt++) {
      try {
        const controller = new AbortController();
        const timeoutId = setTimeout(() => controller.abort(), this.timeout);

        const response = await fetch(url.toString(), {
          method,
          signal: controller.signal,
          headers: {
            'Accept': 'application/json',
            'Content-Type': 'application/json',
          },
        });

        clearTimeout(timeoutId);

        if (!response.ok) {
          const errorText = await response.text();
          let errorMessage = `HTTP ${response.status}: ${response.statusText}`;

          try {
            const errorData = JSON.parse(errorText);
            errorMessage = errorData.detail || errorData.message || errorMessage;
          } catch {
            // Use default message if JSON parsing fails
          }

          if (response.status >= 500 && attempt < this.maxRetries) {
            // Retry server errors
            await new Promise(resolve => setTimeout(resolve, Math.pow(2, attempt) * 1000));
            continue;
          }

          throw new ApiClientError(errorMessage, response.status);
        }

        return await response.json();

      } catch (error) {
        lastError = error as Error;

        if (error instanceof ApiClientError) {
          throw error;
        }

        if (error instanceof Error && error.name === 'AbortError') {
          if (attempt === this.maxRetries) {
            throw new ApiClientTimeoutError(`Request timed out after ${this.maxRetries} retries`);
          }
        } else if (attempt === this.maxRetries) {
          throw new ApiClientError(`Request failed: ${error instanceof Error ? error.message : 'Unknown error'}`);
        }

        // Exponential backoff
        await new Promise(resolve => setTimeout(resolve, Math.pow(2, attempt) * 1000));
      }
    }

    throw lastError!;
  }

  /**
   * Search teams with basic pagination
   */
  async searchTeams(params: TeamSearchParams = {}, useCache = true): Promise<TeamsPaginatedResponse> {
    const searchParams = {
      query: params.query,
      sport_id: params.sport_id,
      league_id: params.league_id,
      market: params.market,
      is_active: params.is_active ?? true,
      page: params.page ?? 1,
      page_size: params.page_size ?? 20,
    };

    // Check cache first
    if (useCache) {
      const cached = this.cache.get('/api/teams/search', searchParams);
      if (cached) {
        return cached;
      }
    }

    const result = await this.request<TeamsPaginatedResponse>('GET', '/api/teams/search', searchParams);

    // Cache the result
    if (useCache) {
      this.cache.set('/api/teams/search', searchParams, result);
    }

    return result;
  }

  /**
   * Enhanced team search with highlighting, relevance scoring, and metadata
   */
  async searchTeamsEnhanced(params: TeamSearchParams = {}, useCache = true): Promise<EnhancedTeamsPaginatedResponse> {
    const searchParams = {
      query: params.query,
      sport_id: params.sport_id,
      league_id: params.league_id,
      market: params.market,
      is_active: params.is_active ?? true,
      page: params.page ?? 1,
      page_size: params.page_size ?? 20,
    };

    // Check cache first
    if (useCache) {
      const cached = this.cache.get('/api/teams/search-enhanced', searchParams);
      if (cached) {
        return cached;
      }
    }

    const result = await this.request<EnhancedTeamsPaginatedResponse>('GET', '/api/teams/search-enhanced', searchParams);

    // Cache the result
    if (useCache) {
      this.cache.set('/api/teams/search-enhanced', searchParams, result);
    }

    return result;
  }

  /**
   * Get search suggestions/autocomplete for team search
   */
  async getSearchSuggestions(query: string, limit = 10, useCache = true): Promise<SearchSuggestionsResponse> {
    if (query.length < 1) {
      throw new Error('Query must be at least 1 character long');
    }

    const params = { query, limit };

    // Check cache first
    if (useCache) {
      const cached = this.cache.get('/api/teams/search-suggestions', params);
      if (cached) {
        return cached;
      }
    }

    const result = await this.request<SearchSuggestionsResponse>('GET', '/api/teams/search-suggestions', params);

    // Cache the result
    if (useCache) {
      this.cache.set('/api/teams/search-suggestions', params, result);
    }

    return result;
  }

  /**
   * Debounced team search - cancels previous search if called again quickly
   */
  async searchTeamsDebounced(
    query?: string,
    params: Omit<TeamSearchParams, 'query'> = {},
    debounceMs = 300
  ): Promise<EnhancedTeamsPaginatedResponse | null> {
    // Cancel previous search
    if (this.abortController) {
      this.abortController.abort();
    }

    this.abortController = new AbortController();

    try {
      // Wait for debounce period
      await new Promise((resolve, reject) => {
        const timeoutId = setTimeout(resolve, debounceMs);
        this.abortController!.signal.addEventListener('abort', () => {
          clearTimeout(timeoutId);
          reject(new Error('Cancelled'));
        });
      });

      // Perform search
      return await this.searchTeamsEnhanced({ ...params, query });

    } catch (error) {
      if (error instanceof Error && error.message === 'Cancelled') {
        return null; // Search was cancelled
      }
      throw error;
    }
  }

  /**
   * Get all sports
   */
  async getSports(
    includeLeagues = false,
    includeInactive = false,
    useCache = true
  ): Promise<SportResponse[]> {
    const params = {
      include_leagues: includeLeagues,
      include_inactive: includeInactive,
    };

    // Check cache first
    if (useCache) {
      const cached = this.cache.get('/api/sports', params);
      if (cached) {
        return cached;
      }
    }

    const result = await this.request<SportResponse[]>('GET', '/api/sports', params);

    // Cache the result
    if (useCache) {
      this.cache.set('/api/sports', params, result);
    }

    return result;
  }

  /**
   * Clear all cached API results
   */
  clearCache(): void {
    this.cache.clear();
  }

  /**
   * Get cache statistics
   */
  getCacheStats(): { cachedItems: number; ttlMs: number } {
    return this.cache.getStats();
  }

  /**
   * Cancel any ongoing debounced searches
   */
  cancelDebouncedSearch(): void {
    if (this.abortController) {
      this.abortController.abort();
    }
  }
}

// Convenience functions
export function createClient(options?: ApiClientOptions): CornerLeagueApiClient {
  return new CornerLeagueApiClient(options);
}

export async function quickSearch(query: string, params?: Omit<TeamSearchParams, 'query'>): Promise<EnhancedTeamsPaginatedResponse> {
  const client = createClient();
  return await client.searchTeamsEnhanced({ ...params, query });
}

export async function quickSuggestions(query: string, limit?: number): Promise<SearchSuggestionsResponse> {
  const client = createClient();
  return await client.getSearchSuggestions(query, limit);
}

// Export default instance
export const apiClient = createClient();

// Example usage
export async function exampleUsage() {
  const client = createClient();

  try {
    // Enhanced search with highlighting
    const results = await client.searchTeamsEnhanced({ query: 'Chicago' });
    console.log(`Found ${results.total} teams in ${results.search_metadata.response_time_ms.toFixed(1)}ms`);

    for (const team of results.items.slice(0, 3)) {
      console.log(`- ${team.display_name} (relevance: ${team.relevance_score})`);
      for (const match of team.search_matches) {
        console.log(`  ${match.field}: ${match.highlighted}`);
      }
    }

    // Get search suggestions
    const suggestions = await client.getSearchSuggestions('Chi');
    console.log('\nSuggestions for "Chi":');
    for (const suggestion of suggestions.suggestions.slice(0, 5)) {
      console.log(`- ${suggestion.suggestion} (${suggestion.type}) - ${suggestion.team_count} teams`);
    }

  } catch (error) {
    console.error('API Error:', error);
  }
}

File: backend/api_client.py
"""
Typed API client for Corner League Media Backend

This module provides a fully typed Python client for the backend API,
with support for enhanced search functionality, caching, and error handling.
"""

import asyncio
import time
from typing import List, Optional, Dict, Any, Union
from uuid import UUID
from datetime import datetime, timedelta
import httpx
from pydantic import BaseModel, ValidationError

# Import the schemas for type safety
try:
    from backend.api.schemas.sports import (
        TeamsPaginatedResponse,
        EnhancedTeamsPaginatedResponse,
        SearchSuggestionsResponse,
        TeamSearchParams,
        SportResponse,
        LeagueResponse
    )
except ImportError:
    # Fallback schemas for standalone usage
    from typing import Any
    TeamsPaginatedResponse = Dict[str, Any]
    EnhancedTeamsPaginatedResponse = Dict[str, Any]
    SearchSuggestionsResponse = Dict[str, Any]
    TeamSearchParams = Dict[str, Any]
    SportResponse = Dict[str, Any]
    LeagueResponse = Dict[str, Any]


class ApiClientError(Exception):
    """Base exception for API client errors"""
    pass


class ApiClientTimeoutError(ApiClientError):
    """Raised when API request times out"""
    pass


class ApiClientValidationError(ApiClientError):
    """Raised when API response validation fails"""
    pass


class SearchCache:
    """Simple in-memory cache for search results"""

    def __init__(self, ttl_seconds: int = 300):  # 5 minutes default
        self.cache: Dict[str, Dict[str, Any]] = {}
        self.ttl_seconds = ttl_seconds

    def _get_cache_key(self, endpoint: str, params: Dict[str, Any]) -> str:
        """Generate cache key from endpoint and parameters"""
        param_str = "&".join(f"{k}={v}" for k, v in sorted(params.items()) if v is not None)
        return f"{endpoint}?{param_str}"

    def get(self, endpoint: str, params: Dict[str, Any]) -> Optional[Any]:
        """Get cached result if still valid"""
        key = self._get_cache_key(endpoint, params)
        if key in self.cache:
            cached_item = self.cache[key]
            if datetime.utcnow() < cached_item['expires']:
                return cached_item['data']
            else:
                del self.cache[key]
        return None

    def set(self, endpoint: str, params: Dict[str, Any], data: Any) -> None:
        """Cache the result with TTL"""
        key = self._get_cache_key(endpoint, params)
        self.cache[key] = {
            'data': data,
            'expires': datetime.utcnow() + timedelta(seconds=self.ttl_seconds)
        }

    def clear(self) -> None:
        """Clear all cached results"""
        self.cache.clear()


class CornerLeagueApiClient:
    """
    Typed API client for Corner League Media Backend

    Features:
    - Fully typed methods with Pydantic validation
    - Request caching for performance
    - Exponential backoff retry logic
    - Comprehensive error handling
    - Search debouncing support
    """

    def __init__(
        self,
        base_url: str = "http://127.0.0.1:8001",
        timeout: float = 30.0,
        cache_ttl: int = 300,
        max_retries: int = 3
    ):
        self.base_url = base_url.rstrip('/')
        self.timeout = timeout
        self.max_retries = max_retries
        self.cache = SearchCache(ttl_seconds=cache_ttl)

        # Create async HTTP client
        self.client = httpx.AsyncClient(
            timeout=httpx.Timeout(timeout),
            limits=httpx.Limits(max_keepalive_connections=10, max_connections=20)
        )

        # Debouncing support
        self._last_search_task: Optional[asyncio.Task] = None
        self._search_debounce_delay = 0.3  # 300ms

    async def __aenter__(self):
        return self

    async def __aexit__(self, exc_type, exc_val, exc_tb):
        await self.close()

    async def close(self):
        """Close the HTTP client"""
        await self.client.aclose()

    async def _request(
        self,
        method: str,
        endpoint: str,
        params: Optional[Dict[str, Any]] = None,
        **kwargs
    ) -> Dict[str, Any]:
        """
        Make HTTP request with retry logic and error handling
        """
        url = f"{self.base_url}{endpoint}"

        # Filter out None and empty string values from params
        if params:
            params = {k: v for k, v in params.items() if v is not None and v != ""}

        for attempt in range(self.max_retries + 1):
            try:
                response = await self.client.request(
                    method=method,
                    url=url,
                    params=params,
                    **kwargs
                )
                response.raise_for_status()
                return response.json()

            except httpx.TimeoutException:
                if attempt == self.max_retries:
                    raise ApiClientTimeoutError(f"Request to {endpoint} timed out after {self.max_retries} retries")
                await asyncio.sleep(2 ** attempt)  # Exponential backoff

            except httpx.HTTPStatusError as e:
                if e.response.status_code < 500 or attempt == self.max_retries:
                    # Don't retry client errors or if we've exhausted retries
                    error_detail = "Unknown error"
                    try:
                        error_data = e.response.json()
                        error_detail = error_data.get('detail', error_data.get('message', str(e)))
                    except:
                        error_detail = str(e)
                    raise ApiClientError(f"API error ({e.response.status_code}): {error_detail}")
                await asyncio.sleep(2 ** attempt)

            except Exception as e:
                if attempt == self.max_retries:
                    raise ApiClientError(f"Unexpected error: {str(e)}")
                await asyncio.sleep(2 ** attempt)

    # Team Search Methods

    async def search_teams(
        self,
        query: Optional[str] = None,
        sport_id: Optional[UUID] = None,
        league_id: Optional[UUID] = None,
        market: Optional[str] = None,
        is_active: Optional[bool] = True,
        page: int = 1,
        page_size: int = 20,
        use_cache: bool = True
    ) -> TeamsPaginatedResponse:
        """
        Search teams with basic pagination

        Args:
            query: Search query for team name or market
            sport_id: Filter by sport UUID
            league_id: Filter by league UUID
            market: Filter by market/city
            is_active: Filter by active status
            page: Page number (1-indexed)
            page_size: Number of items per page
            use_cache: Whether to use cached results

        Returns:
            Paginated team search results
        """
        params = {
            "query": query,
            "sport_id": str(sport_id) if sport_id else None,
            "league_id": str(league_id) if league_id else None,
            "market": market,
            "is_active": is_active,
            "page": page,
            "page_size": page_size
        }

        # Check cache first
        if use_cache:
            cached_result = self.cache.get("/api/teams/search", params)
            if cached_result:
                if hasattr(TeamsPaginatedResponse, 'model_validate') and not isinstance(TeamsPaginatedResponse, type(Dict)):
                    return TeamsPaginatedResponse(**cached_result)
                else:
                    return cached_result

        # Make API request
        data = await self._request("GET", "/api/teams/search", params=params)

        # Validate and parse response
        try:
            if hasattr(TeamsPaginatedResponse, 'model_validate') and not isinstance(TeamsPaginatedResponse, type(Dict)):
                result = TeamsPaginatedResponse(**data)
            else:
                result = data  # Fallback to raw dict
        except (ValidationError, TypeError) as e:
            raise ApiClientValidationError(f"Invalid response format: {str(e)}")

        # Cache the result
        if use_cache:
            self.cache.set("/api/teams/search", params, data)

        return result

    async def search_teams_enhanced(
        self,
        query: Optional[str] = None,
        sport_id: Optional[UUID] = None,
        league_id: Optional[UUID] = None,
        market: Optional[str] = None,
        is_active: Optional[bool] = True,
        page: int = 1,
        page_size: int = 20,
        use_cache: bool = True
    ) -> EnhancedTeamsPaginatedResponse:
        """
        Enhanced team search with highlighting, relevance scoring, and metadata

        Args:
            query: Search query for team name or market
            sport_id: Filter by sport UUID
            league_id: Filter by league UUID
            market: Filter by market/city
            is_active: Filter by active status
            page: Page number (1-indexed)
            page_size: Number of items per page
            use_cache: Whether to use cached results

        Returns:
            Enhanced paginated team search results with metadata
        """
        params = {
            "query": query,
            "sport_id": str(sport_id) if sport_id else None,
            "league_id": str(league_id) if league_id else None,
            "market": market,
            "is_active": is_active,
            "page": page,
            "page_size": page_size
        }

        # Check cache first
        if use_cache:
            cached_result = self.cache.get("/api/teams/search-enhanced", params)
            if cached_result:
                if hasattr(EnhancedTeamsPaginatedResponse, 'model_validate') and not isinstance(EnhancedTeamsPaginatedResponse, type(Dict)):
                    return EnhancedTeamsPaginatedResponse(**cached_result)
                else:
                    return cached_result

        # Make API request
        data = await self._request("GET", "/api/teams/search-enhanced", params=params)

        # Validate and parse response
        try:
            if hasattr(EnhancedTeamsPaginatedResponse, 'model_validate') and not isinstance(EnhancedTeamsPaginatedResponse, type(Dict)):
                result = EnhancedTeamsPaginatedResponse(**data)
            else:
                result = data  # Fallback to raw dict
        except (ValidationError, TypeError) as e:
            raise ApiClientValidationError(f"Invalid response format: {str(e)}")

        # Cache the result
        if use_cache:
            self.cache.set("/api/teams/search-enhanced", params, data)

        return result

    async def get_search_suggestions(
        self,
        query: str,
        limit: int = 10,
        use_cache: bool = True
    ) -> SearchSuggestionsResponse:
        """
        Get search suggestions/autocomplete for team search

        Args:
            query: Search query (minimum 1 character)
            limit: Maximum number of suggestions (1-20)
            use_cache: Whether to use cached results

        Returns:
            Search suggestions with preview teams
        """
        if len(query) < 1:
            raise ValueError("Query must be at least 1 character long")

        params = {"query": query, "limit": limit}

        # Check cache first
        if use_cache:
            cached_result = self.cache.get("/api/teams/search-suggestions", params)
            if cached_result:
                if hasattr(SearchSuggestionsResponse, 'model_validate') and not isinstance(SearchSuggestionsResponse, type(Dict)):
                    return SearchSuggestionsResponse(**cached_result)
                else:
                    return cached_result

        # Make API request
        data = await self._request("GET", "/api/teams/search-suggestions", params=params)

        # Validate and parse response
        try:
            if hasattr(SearchSuggestionsResponse, 'model_validate') and not isinstance(SearchSuggestionsResponse, type(Dict)):
                result = SearchSuggestionsResponse(**data)
            else:
                result = data  # Fallback to raw dict
        except (ValidationError, TypeError) as e:
            raise ApiClientValidationError(f"Invalid response format: {str(e)}")

        # Cache the result
        if use_cache:
            self.cache.set("/api/teams/search-suggestions", params, data)

        return result

    async def search_teams_debounced(
        self,
        query: Optional[str],
        **kwargs
    ) -> Optional[EnhancedTeamsPaginatedResponse]:
        """
        Debounced team search - cancels previous search if called again quickly

        Args:
            query: Search query
            **kwargs: Other search parameters

        Returns:
            Search results or None if cancelled by newer search
        """
        # Cancel previous search task if it exists
        if self._last_search_task and not self._last_search_task.done():
            self._last_search_task.cancel()

        # Create new search task with debounce delay
        async def debounced_search():
            await asyncio.sleep(self._search_debounce_delay)
            return await self.search_teams_enhanced(query=query, **kwargs)

        self._last_search_task = asyncio.create_task(debounced_search())

        try:
            return await self._last_search_task
        except asyncio.CancelledError:
            return None

    # Sports and Leagues Methods

    async def get_sports(
        self,
        include_leagues: bool = False,
        include_inactive: bool = False,
        use_cache: bool = True
    ) -> List[SportResponse]:
        """
        Get all sports

        Args:
            include_leagues: Whether to include leagues for each sport
            include_inactive: Whether to include inactive sports
            use_cache: Whether to use cached results

        Returns:
            List of sports
        """
        params = {
            "include_leagues": include_leagues,
            "include_inactive": include_inactive
        }

        # Check cache first
        if use_cache:
            cached_result = self.cache.get("/api/sports", params)
            if cached_result:
                if hasattr(SportResponse, 'model_validate') and not isinstance(SportResponse, type(Dict)):
                    return [SportResponse(**sport) for sport in cached_result]
                else:
                    return cached_result

        # Make API request
        data = await self._request("GET", "/api/sports", params=params)

        # Validate and parse response
        try:
            if hasattr(SportResponse, 'model_validate') and not isinstance(SportResponse, type(Dict)):
                result = [SportResponse(**sport) for sport in data]
            else:
                result = data  # Fallback to raw list
        except (ValidationError, TypeError) as e:
            raise ApiClientValidationError(f"Invalid response format: {str(e)}")

        # Cache the result
        if use_cache:
            self.cache.set("/api/sports", params, data)

        return result

    # Utility Methods

    def clear_cache(self) -> None:
        """Clear all cached API results"""
        self.cache.clear()

    def get_cache_stats(self) -> Dict[str, int]:
        """Get cache statistics"""
        return {
            "cached_items": len(self.cache.cache),
            "cache_ttl_seconds": self.cache.ttl_seconds
        }


# Convenience functions for synchronous usage

def create_client(**kwargs) -> CornerLeagueApiClient:
    """Create a new API client instance"""
    return CornerLeagueApiClient(**kwargs)


async def quick_search(query: str, **kwargs) -> EnhancedTeamsPaginatedResponse:
    """Quick team search with automatic client management"""
    async with create_client() as client:
        return await client.search_teams_enhanced(query=query, **kwargs)


async def quick_suggestions(query: str, **kwargs) -> SearchSuggestionsResponse:
    """Quick search suggestions with automatic client management"""
    async with create_client() as client:
        return await client.get_search_suggestions(query=query, **kwargs)


# Example usage
if __name__ == "__main__":
    async def example_usage():
        """Example of how to use the API client"""

        # Create client
        async with create_client() as client:
            # Enhanced search with highlighting
            results = await client.search_teams_enhanced(query="Chicago")
            print(f"Found {results.total} teams in {results.search_metadata.response_time_ms:.1f}ms")

            for team in results.items[:3]:
                print(f"- {team.display_name} (relevance: {team.relevance_score})")
                for match in team.search_matches:
                    print(f"  {match.field}: {match.highlighted}")

            # Get search suggestions
            suggestions = await client.get_search_suggestions("Chi")
            print(f"\nSuggestions for 'Chi':")
            for suggestion in suggestions.suggestions[:5]:
                print(f"- {suggestion.suggestion} ({suggestion.type}) - {suggestion.team_count} teams")

    # Run example
    asyncio.run(example_usage())